### DISCUSSION TITLE
Props from getServerSideProps being mutated

### DISCUSSION THREAD

**Discussion by gmcamposano** [vercel/next.js#48159] [Help]
### Summary

Somehow my props from getServerSideProps is mutating whenever i change a state that uses its value as an inital state. I am trying to only update my state variable, and i can see its also updating the data variable that comes from backend. Its not updating it, just mutating the variable.

Why is this?

### Example

https://codesandbox.io/p/sandbox/thirsty-sara-4331i6?file=%2Fpages%2Findex.tsx

**Comment by icyJoseph** [comment]
I think you should take a look at: 

- https://github.com/vercel/next.js/discussions/47135#discussioncomment-5324969

Now onto the harsh reality of JavaScript. Pass by value or pass by reference.

Your Page component receives Props from GetServerSideProps (GSSP) in two ways:

- either on the server as part of the page generation flow, 
- or during SPA transitions on the client, where Next.js makes a request for data to the server, which responds with JSON, which is the result of calling GSSP server side

However there's one issue with data immutability in JavaScript, it is tricky. Let's remove Next.js from the equation, and look at this snippet. 

```js
const obj = {
  foo: {
    bar: [0,1,2,3]
  }
}

const append = (src, item) => {
  src.push(item)
}

const myObj = {...obj}

append(myObj.foo.bar, -1)

console.log(obj.foo.bar) // [ 0, 1, 2, 3, -1 ]
```

We might think that, when we copied the object, by spreading in this case, we can safely mutate said object, but that's not how this operation works, in this case we have a shallow copy.

It is the same with arrays, and other objects:

```js
const arr = [{v: 0}, {v: 1}]

const myArr = [...arr] // or .slice()

myArr[1].v = 1_000

console.log(arr) // [ { v: 0 }, { v: 1000 } ]
```

This is all happening because objects are passed by reference.

> Note that the browser console.log function outputs references, https://giacomocerquone.com/why-does-console-log-hates-me/#other-quicks

In this snippet:

```js
const arr = [0,1,2,3]
const myArr = [...arr]

myArr[0]+= 1000

console.log(arr) // [ 0, 1, 2, 3 ]
console.log(myArr) // [ 1000, 1, 2, 3 ]
```

Numbers are a primitive type of data, that is passed by value, so a copy of them is made, instead of giving you a reference to where that number exists in memory.

So when you do:

```js
  const newFactura = [...factura];
  newFactura[i].producto_id = e.target.value;
  setFactura(newFactura);
```

You are making a new facturas list, yes, but the contents of it are objects, so you have a reference to where in memory they live, and you are changing that! 

So it is nothing to do with Next.js, but "pass by reference" in JavaScript that's messing up with your work here. You might want to consider making a shallow copy of the factura you are changing, but you'd still be mutating the shared factura reference, unless you change:

```js
const [factura, setFactura] = useState(data.factura.slice()); // make a shallow copy
```

Or you could use Immer: https://immerjs.github.io/immer/

This used to be a problem in Redux apps, where for example, an array would be taken from the redux-store, and `.sort()` would be called on it, that mutates the array throughout the app, and introduces very odd looking bugs!

Ask away if its not clear, happy to help!

**Reply by gmcamposano** [reply]
I think in all my years asking things, this has been the best answer i've come across. You made this so clear and makes sense. You got a 5/5 stars right off the bat. 
Oh Javascript, you make some basic things so f*** complicated sometimes.

**Reply by gmcamposano** [reply]
I have bad news :(. I tried the .slice() and even [].concat(data.factura) and the array keeps mutating. 
You can go ahead and try it in the example i demo provided on the post question. On the other hand, using immer worked correctly.

**Reply by icyJoseph** [reply]
Yeah, the thing is that you'd need to copy all the way through, again, you have to understand that `data.factura.slice()`, creates a shallow copy, of that list. 

Perhaps I was not clear, but I meant that you gotta copy at every level, because these copies are shallow, AKA, only one level.

That's what my examples were pointing at how even if you slice at the root level, modifications ripple to every assignment that has a reference. Copying into `myArr` is the same you are trying to do:

```js
const arr = [{v: 0}, {v: 1}]

const myArr = [...arr] // or .slice()

myArr[1].v = 1_000

console.log(arr) // [ { v: 0 }, { v: 1000 } ] 
```

**Reply by icyJoseph** [reply]
So for example:

```js
const [factura, setFactura] = useState(data.factura.slice());

// later on the JSX handlers
  onChange={(e) => {
        const newFactura = [...factura];
                        
        const update = { ...newFactura[i] };
        update.cantidad = Number(e.target.value);
        newFactura[i] = update;
                        
       setFactura(newFactura);
    }}
```

I haven't checked all cases stuff, but roughly like this: https://codesandbox.io/p/sandbox/priceless-lederberg-2216op?file=%2Fpages%2Findex.tsx
